int32 data
float32 time